### PR TITLE
Improvement/cldsrv 724 versionning related functional tests

### PR DESCRIPTION
**Pull request by benzekrimaha**
Issue: CLDSRV-724

### CODE REVIEW

File: tests/functional/aws-node-sdk/test/versioning/listObjectMasterVersions.js [hunk 1]
```diff
@@ -81,53 +85,45 @@ describe('listObject - Delimiter master', function testSuite() {
...
-                                const headers = this.httpResponse.headers;
-                                assert.strictEqual(
-                                    headers['x-amz-delete-marker'], 'true');
-                                return next(err);
-                            });
-                        }
-                        return s3.putObject({
+        it('put objects inside bucket', async () => {
+            for (const obj of objects) {
+                // Handle versioning state changes
```
Comment (ghost) [inline tests/functional/aws-node-sdk/test/versioning/listObjectMasterVersions.js]:
```suggestion
```

File: tests/functional/aws-node-sdk/test/versioning/listObjectMasterVersions.js [hunk 2]
```diff
@@ -81,53 +85,45 @@ describe('listObject - Delimiter master', function testSuite() {
...
+                        Bucket: bucket,
+                        VersioningConfiguration: {
+                            Status: 'Suspended',
+                        },
+                    };
+                    versioning = false;
+                    await s3.send(new PutBucketVersioningCommand(params));
+                }
+
+                // Handle object operations
```
Comment (ghost) [inline tests/functional/aws-node-sdk/test/versioning/listObjectMasterVersions.js]:
```suggestion
```

File: tests/functional/aws-node-sdk/test/versioning/listObjectMasterVersions.js [hunk 3]
```diff
@@ -81,53 +85,45 @@ describe('listObject - Delimiter master', function testSuite() {
...
+                if (obj.value === null) {
+                    // For delete operations, we need to check the response headers
```
Comment (ghost) [inline tests/functional/aws-node-sdk/test/versioning/listObjectMasterVersions.js]:
```suggestion
```

File: tests/functional/aws-node-sdk/test/versioning/listObjectVersions.js [hunk 4]
```diff
@@ -1,408 +1,238 @@
...
+
+        afterEach(done => {
+            removeAllVersions({ Bucket: bucket }, done);
+        });
+        
+
+        it('should list object versions', async () => {
+            const key = 'key';
+            
+            // Put first version
```
Comment (ghost) [inline tests/functional/aws-node-sdk/test/versioning/listObjectVersions.js]:
```suggestion
```

File: tests/functional/aws-node-sdk/test/versioning/listObjectVersions.js [hunk 5]
```diff
@@ -1,408 +1,238 @@
...
+            const putResult1 = await s3.send(new PutObjectCommand({
+                Bucket: bucket,
+                Key: key,
+                Body: 'body1',
+            }));
+            const versionId1 = putResult1.VersionId;
+
+            // Put second version
```
Comment (ghost) [inline tests/functional/aws-node-sdk/test/versioning/listObjectVersions.js]:
```suggestion
```

File: tests/functional/aws-node-sdk/test/versioning/listObjectVersions.js [hunk 6]
```diff
@@ -1,408 +1,238 @@
...
+            const putResult2 = await s3.send(new PutObjectCommand({
+                Bucket: bucket,
+                Key: key,
+                Body: 'body2',
+            }));
+            const versionId2 = putResult2.VersionId;
+
+            // List versions
```
Comment (ghost) [inline tests/functional/aws-node-sdk/test/versioning/listObjectVersions.js]:
```suggestion
```

File: tests/functional/aws-node-sdk/test/versioning/listObjectVersions.js [hunk 7]
```diff
@@ -1,408 +1,238 @@
...
+            const key = 'key';
+            
+            // Put first version
+            await s3.send(new PutObjectCommand({
+                Bucket: bucket,
+                Key: key,
+                Body: 'body1',
+            }));
+
+            // Delete object (creates delete marker)
```
Comment (ghost) [inline tests/functional/aws-node-sdk/test/versioning/listObjectVersions.js]:
```suggestion
```

(same for other comments in the file)

File: tests/functional/aws-node-sdk/test/versioning/listObjectVersions.js [hunk 8]
```diff
@@ -1,408 +1,238 @@
...
-        let versioning = false;
-
-        const objects = [
-            { name: 'notes/summer/august/1.txt', value: 'foo', isNull: true },
-            { name: 'notes/year.txt', value: 'foo', isNull: true },
-            { name: 'notes/yore.rs', value: 'foo', isNull: true },
-            { name: 'notes/zaphod/Beeblebrox.txt', value: 'foo', isNull: true },
-            { name: 'Pâtisserie=中文-español-English', value: 'foo' },
-            { name: 'Pâtisserie=中文-español-English', value: 'bar' },
-            { name: 'notes/spring/1.txt', value: 'qux' },
```
Comment (ghost) [inline tests/functional/aws-node-sdk/test/versioning/listObjectVersions.js:61]: are we stopping supporting object names like these? There is an advantage of keeping the special characters, as we do listing, it's to ensure it works well even when the object names are uncommon

File: tests/functional/aws-node-sdk/test/versioning/multiObjectDelete.js [hunk 9]
```diff
@@ -46,141 +42,136 @@ describe('Multi-Object Versioning Delete Success', function success() {
...
+
+            const objects = [];
+            for (let i = 1; i < 1001; i++) {
+                objects.push(`${key}${i}`);
+            }
 
-        afterEach(async () => {
-            await removeAllVersions({ Bucket: bucketName });
-            await s3.send(new DeleteBucketCommand({ Bucket: bucketName }));
+            // Create objects in batches of 20 concurrently
```
Comment (ghost) [inline tests/functional/aws-node-sdk/test/versioning/multiObjectDelete.js:59]:
```suggestion
```

**Review by ghost — COMMENTED** [review]
just one question around listing and special characters, otherwise lgtm

File: tests/functional/aws-node-sdk/lib/utility/customS3Request.js [hunk 10]
```diff
@@ -1,43 +1,59 @@
...
+const { S3Client } = require('@aws-sdk/client-s3');
+const { HttpRequest } = require('@smithy/protocol-http');
 const querystring = require('querystring');
 
 const getConfig = require('../../test/support/config');
 
-const config = getConfig('default', { signatureVersion: 'v4' });
-const s3 = new S3(config);
+const config = getConfig('default');
+
```
Comment (DarkIsDude) [inline tests/functional/aws-node-sdk/lib/utility/customS3Request.js]:
```suggestion
```

nits

File: tests/functional/aws-node-sdk/lib/utility/customS3Request.js [hunk 11]
```diff
@@ -1,43 +1,59 @@
...
+        headers: newHeaders,
+        query: newQuery,
     });
-    request.on('error', err => {
+
+    return next({ ...args, request: newReq });
+};
+
+async function customS3Request(CommandClass, params, buildParams) {
+        const customS3 = new S3Client({ ...config });
```
Comment (DarkIsDude) [inline tests/functional/aws-node-sdk/lib/utility/customS3Request.js]: indentation issue

File: tests/functional/aws-node-sdk/lib/utility/versioning-util.js [hunk 12]
```diff
@@ -28,15 +28,14 @@ function _deleteVersionList(versionList, bucket, callback) {
     return s3Client.send(new DeleteObjectsCommand(params)).then(() => callback()).catch(err => callback(err));
 }
 
-function checkOneVersion(s3, bucket, versionId, callback) {
+function checkOneVersion(s3, bucket, versionId) {
     return s3Client.send(new ListObjectVersionsCommand({ Bucket: bucket })).then(data => {
```
Comment (DarkIsDude) [inline tests/functional/aws-node-sdk/lib/utility/versioning-util.js]:
```suggestion
async function checkOneVersion(s3, bucket, versionId) {
return await s3Client.send(new ListObjectVersionsCommand({ Bucket: bucket })).then(data => {
```

This will keep the `checkOneVersion` in the stack trace. It's a "good" practice for async/await. But nits

File: tests/functional/aws-node-sdk/test/versioning/bucketDelete.js [hunk 13]
```diff
@@ -25,72 +27,70 @@ describe('aws-node-sdk test delete bucket', () => {
...
+                Bucket: bucketName,
+                VersioningConfiguration: {
+                    Status: 'Enabled',
+                },
+            }));
         });
 
         // empty and delete bucket after testing if bucket exists
-        afterEach(done => {
+         afterEach(done => {
```
Comment (DarkIsDude) [inline tests/functional/aws-node-sdk/test/versioning/bucketDelete.js]:
```suggestion
        afterEach(done => {
```

indentation ? Weird that the CI pass 😬

File: tests/functional/aws-node-sdk/test/versioning/objectCopy.js [hunk 14]
```diff
@@ -53,20 +70,19 @@ function checkNoError(err) {
...
 }
 
 function dateFromNow(diff) {
     const d = new Date();
     d.setHours(d.getHours() + diff);
-    return d.toISOString();
+    return d;
 }
 
 function dateConvert(d) {
```
Comment (DarkIsDude) [inline tests/functional/aws-node-sdk/test/versioning/objectCopy.js:82]: This helper seems really useful ?

File: tests/functional/aws-node-sdk/test/versioning/objectDelete.js [hunk 15]
```diff
@@ -18,107 +30,77 @@ const nonExistingId = process.env.AWS_ON_AIR ?
...
-                        return callback();
-                    }),
-            ], done);
+        async () => {
+            await s3.send(new PutBucketVersioningCommand({
+                Bucket: bucket,
+                VersioningConfiguration: versioningSuspended,
+            }));
+
+            // List versions to check null version exists
```
Comment (DarkIsDude) [inline tests/functional/aws-node-sdk/test/versioning/objectDelete.js]: I would delete comments as they just describe the test (same before)

**Review by DarkIsDude — APPROVED** [review]
LGTM. Just some nits but good 👏

**Review by DarkIsDude — CHANGES_REQUESTED** [review]
test

**Review by DarkIsDude — APPROVED** [review]

File: tests/functional/aws-node-sdk/test/versioning/listObjectMasterVersions.js [hunk 16]
```diff
@@ -37,21 +44,18 @@ describe('listObject - Delimiter master', function testSuite() {
...
 
         // setup test
-        before(done => {
-            s3.createBucket({ Bucket: bucket }, done);
+        before(async () => {
+            await s3.send(new CreateBucketCommand({ Bucket: bucket }));
         });
 
-        // delete bucket after testing
+
```
Comment (SylvainSenechal) [inline tests/functional/aws-node-sdk/test/versioning/listObjectMasterVersions.js]:
```suggestion
```

File: tests/functional/aws-node-sdk/test/versioning/listObjectVersions.js [hunk 17]
```diff
@@ -1,408 +1,238 @@
...
-                            test.nextKeyMarker);
-                        if (!test.nextVersionIdMarker) {
-                            // eslint-disable-next-line no-param-reassign
-                            test.nextVersionIdMarker = {};
-                        }
-                        assert.strictEqual(res.NextVersionIdMarker,
-                            test.nextVersionIdMarker.versionId);
-                        return done();
-                    });
+        it('should handle pagination with MaxKeys', async () => {
```
Comment (SylvainSenechal) [inline tests/functional/aws-node-sdk/test/versioning/listObjectVersions.js]: Kinda agree with William's comment, I feel like some of the tests got removed, and we are missing coverage for some corner case / special characters ?

File: tests/functional/aws-node-sdk/test/versioning/objectCopy.js [hunk 18]
```diff
@@ -103,1057 +116,779 @@ describe('Object Version Copy', () => {
...
+            const copyRes = await s3.send(new CopyObjectCommand({ Bucket: destBucketName, Key: destObjName,
+                CopySource: copySource }));
+            assert.strictEqual(copyRes.CopyObjectResult.ETag, emptyFileETag);
+            const getRes = await s3.send(new GetObjectCommand({ Bucket: destBucketName,
+                Key: destObjName }));
+            assert.deepStrictEqual(getRes.Metadata, originalMetadata);
+            assert.strictEqual(getRes.ETag, emptyFileETag);
         });
 
-        // TODO: remove (or update to use different location constraint) in CLDSRV-639
```
Comment (SylvainSenechal) [inline tests/functional/aws-node-sdk/test/versioning/objectCopy.js:383]: Feels like this should be kept, cldsrv 639 is 7 months old last updated in october

File: tests/functional/aws-node-sdk/test/versioning/objectCopy.js [hunk 19]
```diff
@@ -103,1057 +116,779 @@ describe('Object Version Copy', () => {
...
+                    CopySourceIfUnmodifiedSince: dateFromNow(1) });
+                assert.fail('Expected error');
+            } catch (err) {
                 checkError(err, 'PreconditionFailed');
-                done();
-            });
+            }
         });
 
-        // Skipping this test, because real AWS does not provide error as
```
Comment (SylvainSenechal) [inline tests/functional/aws-node-sdk/test/versioning/objectCopy.js:785]: should be kept

File: tests/functional/aws-node-sdk/test/versioning/objectCopy.js [hunk 20]
```diff
@@ -103,1057 +116,779 @@ describe('Object Version Copy', () => {
...
                 CopySourceIfNoneMatch: 'non-matching',
-                CopySourceIfModifiedSince: dateFromNow(-1),
-            }, err => {
-                checkNoError(err);
-                done();
-            });
+                CopySourceIfModifiedSince: dateFromNow(-1) });
         });
 
-        // Skipping this test, because real AWS does not provide error as
```
Comment (SylvainSenechal) [inline tests/functional/aws-node-sdk/test/versioning/objectCopy.js:848]: same

File: tests/functional/aws-node-sdk/test/versioning/objectPut.js [hunk 21]
```diff
@@ -1,575 +1,419 @@
...
-        'versionId query', done => {
-            const params = { Bucket: bucket, Key: key };
-            const query = { versionId: '' };
-            customS3Request(s3.putObject, params, { query }, err => {
-                assert(err, 'Expected error but did not find one');
-                assert.strictEqual(err.code, 'InvalidArgument');
-                assert.strictEqual(err.statusCode, 400);
-                done();
-            });
+        afterEach(done => {
```
Comment (SylvainSenechal) [inline tests/functional/aws-node-sdk/test/versioning/objectPut.js]: arent we doing removeAllVersions twice here ? We have after() and afterEach(), we used to have only afterEach() 🤔

File: tests/functional/aws-node-sdk/test/versioning/objectPut.js [hunk 22]
```diff
@@ -1,575 +1,419 @@
...
+                Key: key,
+                Body: 'body',
+            }));
+
+            assert(putResult.VersionId);
+            assert.notEqual(putResult.VersionId, 'null');
+            assert.strictEqual(putResult.$metadata.httpStatusCode, 200);
         });
 
-        describe('on a version-enabled bucket', () => {
```
Comment (SylvainSenechal) [inline tests/functional/aws-node-sdk/test/versioning/objectPut.js:104]: It's a bit hard to follow the diff, the test names were changes, the old ones had a putBucketVersioning command, but I can't find it with the new tests

File: tests/functional/aws-node-sdk/test/versioning/objectPut.js [hunk 23]
```diff
@@ -1,575 +1,419 @@
...
-                        _assertNoError(err, 'getting null version');
-                        assert.strictEqual(nullVerData.ETag, eTags[0]);
-                        callback();
-                    }),
-                ], done);
-            });
+                    VersionId: putResults[i].VersionId,
+                }));
 
-            it('should create a bunch of objects and their versions', done => {
```
Comment (SylvainSenechal) [inline tests/functional/aws-node-sdk/test/versioning/objectPut.js]: This test was written twice in the old code, right ?

**Review by SylvainSenechal — CHANGES_REQUESTED** [review]
Can you address the comments : there are 2 files where the tests looks quite different from the original one (one file where it seems some tests are missing, another one where the tests are renamed/looks differents). It might not be a problem, but I wonder why it was changed this way

File: tests/functional/aws-node-sdk/test/versioning/objectCopy.js (hunk 14 above)
Comment (benzekrimaha) [reply tests/functional/aws-node-sdk/test/versioning/objectCopy.js:82]: keeping it to avoid additional changes , only adapting it to suite the new format required by the sdk v3

File: tests/functional/aws-node-sdk/test/versioning/objectPut.js (hunk 21 above)
Comment (benzekrimaha) [reply tests/functional/aws-node-sdk/test/versioning/objectPut.js]: doesn't hurt to keep it as an additionnal security level

**Review by SylvainSenechal — APPROVED** [review]
